### PR TITLE
Added check before dropping well column

### PR DESCRIPTION
**Pull request by kenibrewer**
# Description

Simple bugfix for #240  where a missing check for identical column names results in an inappropriately dropped column.

## What is the nature of your change?

- [x] Bug fix (fixes an issue).
- [ ] Enhancement (adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] This change requires a documentation update.

# Checklist

Please ensure that all boxes are checked before indicating that a pull request is ready for review.

- [x] I have read the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] New and existing unit tests pass locally with my changes.

### CODE REVIEW

**Comment by gwaybio** [comment]
Thanks for your contribution @kenibrewer - can you add a small test that would have failed the previous functionality that this contribution now fixes?

**Comment by gwaybio** [comment]
Actually, maybe the simpler solution would be to drop the expression `.drop(join_on[0], axis="columns"`

But, then we run into the issue with auto suffixes changing both column names. We'd have to add something like: `suffixes=("_platemap", "")`

What do you think?

**Comment by kenibrewer** [comment]
@gwaybio I added an example to the related issue page that illustrates the problem. I also added a test_case to test_annotate.py that crashes with the old code and works with the new code. In the process, I also discovered that a failure to work on fresh dataframe copies in individual tests was causing operations in one test to affect the others. So I fixed that too.

> Actually, maybe the simpler solution would be to drop the expression `.drop(join_on[0], axis="columns"`
> 
> But, then we run into the issue when `join_on[0] == join_on[1]` with auto suffixes changing both column names. We'd have to add something like: `suffixes=("_platemap", "")`
> 
> What do you think?

I think there is value to eliminating redundant information by having dropping the platemap version of the well column name. So I think we should still drop the column when appropriate. However, even if we drop the expression `.drop(join_on[0], axis="columns"`, we won't get any additional columns with auto-suffixes than we would anyways. 

Addressing the auto_suffix problem separately, I actually do really like the idea of adding the `suffixes=("_platemap", "")` parameter regardless. An additional nice feature would be if after the merge any new '_platemap' columns are checked against their matching data column and are dropped if they are identical.

File: pycytominer/tests/test_annotate.py
Comment (gwaybio) [inline pycytominer/tests/test_annotate.py:40]: I'm a bit confused why this is necessary - our tests were passing before.

I think you can remove the copies and revert the changed lines back to the original data.

File: pycytominer/tests/test_annotate.py
Comment (gwaybio) [inline pycytominer/tests/test_annotate.py:63]: delete this line please

File: pycytominer/tests/test_annotate.py
Comment (gwaybio) [inline pycytominer/tests/test_annotate.py:62]:
```suggestion
    platemap_df = PLATEMAP_DF.copy(deep=True)
    platemap_df = platemap_df.rename(columns={"well_position":"well"})
```

To confirm that any modifications that you perform on the copy doesn't bleed through to the original `PLATEMAP_DF`

File: pycytominer/tests/test_annotate.py
Comment (gwaybio) [inline pycytominer/tests/test_annotate.py:75]: These two entries are different - did you mean to make them the same? (case sensitive)

File: pycytominer/tests/test_annotate.py
Comment (gwaybio) [inline pycytominer/tests/test_annotate.py:81]: I am confused about why `test_annotate_output` needed to be modified.

File: pycytominer/tests/test_annotate.py
Comment (gwaybio) [inline pycytominer/tests/test_annotate.py:106]: Also, why did this need to be modified? (`test_annotate_output_compress`)

The tests were passing previously

**Review by gwaybio — CHANGES_REQUESTED** [review]
A couple discussion questions and required changes prior to review. Thanks again for your contribution!

File: pycytominer/tests/test_annotate.py
Comment (kenibrewer) [reply pycytominer/tests/test_annotate.py:40]: In the live version of the code, tests are passing inappropriately because PLATEMAP_DF columns being renamed during the execution of `test_annotate()` is cancelled out by the inclusion of incorrect arguments in `test_annotate_output()` and `test_annotate_output_compress()`

### PLATEMAP_DF column names are modified during`test_annotate()`

Because `test_annotate()` uses the default argument `add_metadata_id_to_platemap=True`, the platemap column names are being modified during execution. This modification occurs [here](https://github.com/cytomining/pycytominer/blob/ea9873ddd3b7d9b7072cf7fc55d582012cf6d315/pycytominer/annotate.py#L78) in the load_platemap function. Calling `platemap_df = PLATEMAP_DF.copy()` and using the copied dataframe is necessary to prevent the column renaming from influencing other tests. 

### Incorrect Inputs
 Both `test_annotate_output()` and `test_annotate_output_compress()` contain the argument `add_metadata_id_to_platemap=False`. This means that for both of them, the lines `join_on=["Metadata_well_position", "Metadata_Well"]` should be changed to `join_on=["well_position", "Metadata_Well"]`  because "well_position" is the appropriate column name. Correcting these incorrect arguments will cause those tests to fail unless we modify `test_annotate()` with the line of code being discussed.. 





The reason that these tests were

File: pycytominer/tests/test_annotate.py
Comment (kenibrewer) [reply pycytominer/tests/test_annotate.py:62]: Because I am only modifying column names of PLATEMAP_DF for this new test and am not changing any of the underlying data, `copy(deep=True)` is not necessary. 

Here is chart that I believe summarizes when which type of copy is required: 
|                        | No changes to Columns | Changes to Columns |
|------------------------|-----------------------|--------------------|
| **No changes to Data** | No copy needed        | .copy(deep=False)  |
| **Changes to Data**    | .copy(deep=True)      | .copy(deep=True)   |

File: pycytominer/tests/test_annotate.py
Comment (kenibrewer) [reply pycytominer/tests/test_annotate.py:81]: More detailed explanation above, but in short `test_annotate_output` is operating on a modified version of `PLATEMAP_DF` and is only passing because it is being given the wrong arguments.

File: pycytominer/tests/test_annotate.py
Comment (kenibrewer) [reply pycytominer/tests/test_annotate.py:106]: More detailed explanation above, but in short `test_annotate_output_compress` is operating on a version of `PLATEMAP_DF` that has renamed columns and is only passing because it is being given the wrong arguments.

File: pycytominer/tests/test_annotate.py
Comment (kenibrewer) [reply pycytominer/tests/test_annotate.py:81]: The use of `platemap_df = PLATEMAP_DF.copy()` is not strictly necessary, but I think it maybe worth adopting that structure in all of the tests for consistency sake.

File: pycytominer/tests/test_annotate.py
Comment (kenibrewer) [reply pycytominer/tests/test_annotate.py:106]: As mentioned above, the use of `platemap_df = PLATEMAP_DF.copy()` is not strictly necessary for this test, but I think it maybe worth adopting that structure in all of the tests for consistency sake.

File: pycytominer/tests/test_annotate.py
Comment (kenibrewer) [reply pycytominer/tests/test_annotate.py:62]: There may be some value to using `copy(deep=True)` even though it isn't necessary in this case as a model for future test writers who want to need to operate on modified versions of `PLATEMAP_DF`.

File: pycytominer/tests/test_annotate.py
Comment (gwaybio) [reply pycytominer/tests/test_annotate.py:40]: Amazing find @kenibrewer ! Thank you for bringing this to our attention.

I think this issue is worth resolving at the source. Can you add a data frame `.copy(deep=True)` to the `load_platemap()` function [here](https://github.com/cytomining/pycytominer/blob/ea9873ddd3b7d9b7072cf7fc55d582012cf6d315/pycytominer/cyto_utils/load.py#L78)?

Maybe something like:

```python
    if add_metadata_id:
        platemap_df = platemap.copy(deep=True)
        platemap_df.columns = [
            f"Metadata_{x}" if not x.startswith("Metadata_") else x
            for x in platemap_df.columns
        ]
    return platemap_df
```

(you'll need to modify `platemap_df = pd.read_csv(platemap, sep=delim)` a couple lines above as well.

I think that it's worth fixing this bug than fixing tests that work around the bug. (you'll also likely have to update tests that were inappropriately passing!)

Thanks again!

**Comment by gwaybio** [comment]
@kenibrewer - please correct me if I'm wrong, but I think we're waiting on your reply to [this comment](https://github.com/cytomining/pycytominer/pull/241#discussion_r1028571680) prior to moving forward with merging this PR. Thanks again!

File: pycytominer/tests/test_annotate.py
Comment (kenibrewer) [reply pycytominer/tests/test_annotate.py:40]: Hi @gwaybio ,
I think you are correct about it being better to solve the problem by addressing the root cause of the bug which is the failure to use `.copy(deep=True)`.  I'll make the appropriate adjustments.

**Comment by kenibrewer** [comment]
Hi @gwaybio, 
You are correct that you are waiting on me to submit some additional changes prior to merging. I had some other priorities the past few weeks but have some time to work on this now. 
Related to this bug. I've think I've discovered some other bugs related to missing `.copy(deep=True)` in other pycytominer functions. I'll submit those as a separate issue though.

**Comment by kenibrewer** [comment]
Revisiting your [suggestion about adding suffixes](https://github.com/cytomining/pycytominer/pull/241#issuecomment-1315459163), I think it would be good to add `suffixes = ["", "platemap"]` to the platemap merge and `suffixes = ["", "external"] to the external metadata merge. Both of those changes would result in much more readable and predictable outputs than the default "_x" and "_y" suffixes. 

The one problem with this change, however, is that it would likely break pipelines/code (including a few of my own projects), that already include logic around renaming/handling the "_x" and "_y" columns that are generated by the current codebase. I still think the change is worthwhile, but I'm not sure how you handle potentially breaking changes.

File: pycytominer/tests/test_annotate.py
Comment (kenibrewer) [reply pycytominer/tests/test_annotate.py]: Thanks. Fixed.

File: pycytominer/tests/test_annotate.py
Comment (kenibrewer) [reply pycytominer/tests/test_annotate.py]: Done

**Comment by gwaybio** [comment]
Thanks @kenibrewer - thanks for following up and I'm excited you're able to work on this PR again. I'll respond to a couple of your thoughts and then provide a code review.

> I've think I've discovered some other bugs related to missing .copy(deep=True) in other pycytominer functions. I'll submit those as a separate issue though.

Fantastic - thanks! Once you create a new issue, please link this PR

> I think it would be good to add suffixes = ["", "platemap"] to the platemap merge and suffixes = ["", "external"] to the external metadata merge.

Sounds good. My sense is that this change belongs in a separate PR. Let's focus on dropping the well column (and fixing any immediate bugs that arise from this closer investigation) here, and save this potentially breaking change for a subsequent PR.

> I'm not sure how you handle potentially breaking changes.

We haven't had a hard and fast policy, but we typically mint a new release version which will kickoff a new pypi and conda build. If you're using a specific pycytominer version (or github hash) in your previous pipelines, then this should have no impact on past work other than needing to modify code for future projects if new pycytominer versions are good enough to warrant a version upgrade.

File: pycytominer/cyto_utils/load.py
Comment (gwaybio) [inline pycytominer/cyto_utils/load.py:78]:
```suggestion
        # Setting platemap to a copy to prevent column name changes from back-propagating
```

File: pycytominer/tests/test_annotate.py
Comment (gwaybio) [inline pycytominer/tests/test_annotate.py:58]: this is really minor, but can you name `platemap_df` something else more descriptive? (just to increase readability)

**Review by gwaybio — COMMENTED** [review]
Looks good! I'll kick off tests now. Please review my minor suggestions and then, once the tests pass, I'll check in with you and merge.

Thanks again for your contribution!

File: pycytominer/cyto_utils/load.py
Comment (kenibrewer) [reply pycytominer/cyto_utils/load.py:78]: Fixed

File: pycytominer/tests/test_annotate.py
Comment (kenibrewer) [reply pycytominer/tests/test_annotate.py]: Fixed.

**Comment by gwaybio** [comment]
Looks like the tests will pass - are we set to merge this in @kenibrewer ?

**Comment by kenibrewer** [comment]
@gwaybio Yep. Ready for merge. 🚀

**Review by gwaybio — APPROVED** [review]
🚀

**Comment by gwaybio** [comment]
Thanks again @kenibrewer 

When you get a chance, please create new issues in https://github.com/cytomining/pycytominer/issues. I believe they involved 1) `copy()` based bugs and 2) specific merge suffixes. I don't want to lose track of these

**Pull request merged**
